cmd/soroban-cli/src/self_outdated_check.rs

```diff
@@ -0,0 +1,159 @@
+use crate::config::self_outdated_check::SelfOutdatedCheck;
+use crate::print::Print;
+use semver::Version;
+use serde::Deserialize;
+use std::error::Error;
+use std::io::IsTerminal;
+use std::time::Duration;
+
+const MINIMUM_CHECK_INTERVAL: Duration = Duration::from_secs(60 * 60 * 24); // 1 day
+const CRATES_IO_API_URL: &str = "https://crates.io/api/v1/crates/";
+const REQUEST_TIMEOUT: Duration = Duration::from_secs(30);
+const NO_UPDATE_CHECK_ENV_VAR: &str = "STELLAR_NO_UPDATE_CHECK";
+
+#[derive(Deserialize)]
+struct CrateResponse {
+    #[serde(rename = "crate")]
+    crate_: Crate,
+}
+
+#[derive(Deserialize)]
+struct Crate {
+    #[serde(rename = "max_stable_version")]
+    max_stable_version: Version,
+    #[serde(rename = "max_version")]
+    max_version: Version, // This is the latest version, including pre-releases
+}
+
+/// Fetch the latest stable version of the crate from crates.io
+fn fetch_latest_crate_info() -> Result<Crate, Box<dyn Error>> {
+    let crate_name = env!("CARGO_PKG_NAME");
+    let url = format!("{CRATES_IO_API_URL}{crate_name}");
+    let response = ureq::get(&url).timeout(REQUEST_TIMEOUT).call()?;
+    let crate_data: CrateResponse = response.into_json()?;
+    Ok(crate_data.crate_)
+}
+
+/// Print a warning if a new version of the CLI is available
+pub fn print_upgrade_prompt() {
+    // We should skip the upgrade check if we're not in a tty environment.
+    if !std::io::stderr().is_terminal() {
+        return;
+    }
+
+    // We should skip the upgrade check if the user has disabled it by setting
+    // the environment variable (STELLAR_NO_UPDATE_CHECK)
+    if std::env::var(NO_UPDATE_CHECK_ENV_VAR).is_ok() {
+        return;
+    }
+
+    let current_version = crate::commands::version::pkg();
+    let print = Print::new(false);
+
+    let mut stats = match SelfOutdatedCheck::load() {
+        Ok(stats) => stats,
+        Err(e) => {
+            print.warnln(format!("Failed to load self outdated check data: {}", e));
```

Comment:
I think we should send this line to the logs instead of to stderr. The update check is supporting functionality, and if it fails, but the user is still able to move ahead with the action they're performing we shouldn't get in the users way with info about the update check imo.

Reply:
Changed to use `println!` for output.

Reply:
Sorry, I should have been more specific, we need to use `tracing::warn` to write to the logs. The logs are a separate stream that can be enabled/disabled separately to the main application output.

Reply:
Thank you for pointing this out. I initially thought there was a log library here, but when I searched for `log`, I didn't find it. 🥲

Reply:
Fixed in 5e3c073a206688e42fbcb049ab12499d2e120993.
It seems that we are currently unable to place `thread::spawn(upgrade_check);` at the very beginning.